plugins/out_azure_kusto/azure_kusto.c

```diff
@@ -18,61 +18,266 @@
-        return -1;
+    flb_plg_debug(ctx->ins, "[flb_azure_imds_get_token] HTTP response status: %d & payload: %.*s", client->resp.status, (int)client->resp.payload_size, client->resp.payload);
+
+    response = flb_sds_create_len(client->resp.payload, client->resp.payload_size);
+    flb_http_client_destroy(client);
+    flb_upstream_conn_release(u_conn);
+    flb_sds_destroy(url);
+
+    return response;
+}
+
+
+/**
+ * azure_kusto_get_oauth2_token - Retrieve an OAuth2 token for Azure Kusto
+ *
+ * This function is responsible for obtaining an OAuth2 access token required
+ * for authenticating requests to Azure Kusto. It supports two methods of
+ * token retrieval: using the Azure Instance Metadata Service (IMDS) or
+ * through a client secret flow.
+ *
+ * Parameters:
+ *   - ctx: A pointer to the flb_azure_kusto context structure, which contains
+ *          configuration and state information necessary for token retrieval.
+ *
+ * Returns:
+ *   - 0 on success, indicating that the token was successfully retrieved and
+ *     stored in the context.
+ *   - -1 on failure, indicating an error occurred during the token retrieval
+ *     process.
+ *
+ * The function performs the following steps:
+ *   1. Logs the start of the token retrieval process.
+ *   2. Checks if IMDS-based token retrieval is enabled:
+ *      - If enabled, it attempts to retrieve a token from the Azure IMDS.
+ *      - Parses the JSON response to extract the access token.
+ *      - Logs and handles errors if token extraction fails.
+ *   3. If IMDS is not used, it falls back to the client secret flow:
+ *      - Clears any existing OAuth2 payload data.
+ *      - Appends necessary parameters to the OAuth2 payload, including
+ *        grant type, scope, client ID, and client secret.
+ *      - Requests an OAuth2 token using the constructed payload.
+ *      - Logs and handles errors if token retrieval fails.
+ *   4. Logs the successful completion of the token retrieval process.
+ *
+ * Note:
+ *   - The function assumes that the context (ctx) is properly initialized
+ *     and contains valid configuration data.
+ *   - The extracted access token is stored in the context's OAuth2 structure
+ *     for subsequent use in authenticated requests.
+ */
+
+static int azure_kusto_get_oauth2_token(struct flb_azure_kusto *ctx)
+{
+    char *token = NULL;
+    int ret;
+    flb_sds_t response = NULL;
+    char *access_token = NULL;
+
+    flb_plg_debug(ctx->ins, "Starting token retrieval process");
+
+    /* Check if IMDS-based token retrieval is needed */
+    if (ctx->use_imds == FLB_TRUE) {
+        flb_plg_debug(ctx->ins, "Using IMDS for token retrieval");
+
+        response = flb_azure_imds_get_token(ctx, ctx->client_id);
+        if (!response) {
+            flb_plg_error(ctx->ins, "Failed to retrieve token from Azure IMDS");
+            return -1;
+        }
+
+        // Parse the JSON response to extract the access token
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#comments-in-the-code